src/lib/paginator/paginator.ts

```diff
@@ -76,26 +76,42 @@ export class MatPaginator extends _MatPaginatorBase implements OnInit, OnDestroy
   @Input()
   get pageIndex(): number { return this._pageIndex; }
   set pageIndex(value: number) {
-    this._pageIndex = Math.max(coerceNumberProperty(value), 0);
-    this._changeDetectorRef.markForCheck();
+    const newPageIndex = Math.max(coerceNumberProperty(value), 0);
+    const oldPageIndex = this._pageIndex;
+
+    if (newPageIndex !== oldPageIndex) {
+      this._pageIndex = newPageIndex;
+      this._emitPageEvent(oldPageIndex);
+      this._changeDetectorRef.markForCheck();
+    }
   }
   _pageIndex: number = 0;
 
   /** The length of the total number of items that are being paginated. Defaulted to 0. */
   @Input()
   get length(): number { return this._length; }
   set length(value: number) {
-    this._length = coerceNumberProperty(value);
-    this._changeDetectorRef.markForCheck();
+    const newLength = coerceNumberProperty(value);
+
+    if (newLength !== this._length) {
+      this._length = newLength;
+      this._emitPageEvent(this.pageIndex);
+      this._changeDetectorRef.markForCheck();
+    }
   }
   _length: number = 0;
 
   /** Number of items to display on a page. By default set to 50. */
   @Input()
   get pageSize(): number { return this._pageSize; }
   set pageSize(value: number) {
-    this._pageSize = Math.max(coerceNumberProperty(value), 0);
-    this._updateDisplayedPageSizeOptions();
+    const newPageSize = Math.max(coerceNumberProperty(value), 0);
+
+    if (newPageSize !== this._pageSize) {
+      this._pageSize = newPageSize;
+      this._updateDisplayedPageSizeOptions();
+      this._emitPageEvent(this.pageIndex);
```

Comment:
Should this also markForCheck?

Reply:
I don't think it needs to since it's going through an `Output`. At least in my testing it wasn't necessary.